### ISSUE TITLE
you did not implement Re-param part......

### ISSUE THREAD

**Issue by busyyang** [YeongHyeon/RepVGG-TF2#1]
Hello, I survey into this repo and I find you just give up the `branch_sub` part when inference.
~~~py
    def repvgg(self, input, ksize, inchannel, outchannel, stride_size=1,  name="", train=False, verbose=False):
        .......

        if (train):
            if (stride_size == 1):
                output = branch_main_bn + branch_sub_bn + input
            else:
                output = branch_main_bn + branch_sub_bn
        else:
            output = branch_main_bn
        output_act = self.customlayers.elu(output)
~~~

But in chapter 3 in original paper, the authors map the `branch_sub_bn ` and `input` part into a 3x3  filter. This is totally wrong way to re-implement RepVGG......

**Comment by YeongHyeon** [comment]
Hello busyyang,

Thank you for your interesting.
Can you elaborate a little bit about what's wrong?
I added the comment to repvgg module for you as below.

``` python
def repvgg(self, input, ksize, inchannel, outchannel, stride_size=1, \
        name="", train=False, verbose=False):

        """
        1. branch_main (branch_main_bn)
                The k x k convolutional layer. 
                As an example, it becomes 3x3 when called.
                This layer is always used when training and inference.
        2. branch_sub_bn (branch_sub_bn)
                The 1 x 1 convolutional layer.
                This layer is used only inference procedure.
        """
        branch_main = self.customlayers.conv2d(input, \
            self.customlayers.get_weight(vshape=[ksize, ksize, inchannel, outchannel], name="%s_main" %(name)), \
            stride_size=stride_size, padding='SAME')
        branch_main_bn = self.customlayers.batch_norm(branch_main, name="%s_main_bn" %(name))

        branch_sub = self.customlayers.conv2d(input, \
            self.customlayers.get_weight(vshape=[1, 1, inchannel, outchannel], name="%s_sub" %(name)), \
            stride_size=stride_size, padding='SAME')
        branch_sub_bn = self.customlayers.batch_norm(branch_sub, name="%s_sub_bn" %(name))

        """
        When training procedure, output is constructed by two or three input.
        In case of stride 1, output is sum of main-branch, sub-branch, and original input.
        In the other case, the case of stride 2, the output is the sum of main-branch and sub-branch only. 
        (cannot add original input by different shape)

        When inference procedure, the output is constructed with only main-branch.
        The above method called 're-parameterization' in chapter 3 of original paper.
        """
        if(train):
            if(stride_size == 1):
                output = branch_main_bn + branch_sub_bn + input
            else:
                output = branch_main_bn + branch_sub_bn
        else:
            output = branch_main_bn
        output_act = self.customlayers.elu(output)

        if(verbose): print(name, output_act.shape)
        return output_act
```

Thank you once more and have a nice day!
:)

**Comment by busyyang** [comment]
Hi YeongHyeon, 
You absolutely finished the training part as the picture you showed. But for inference, not only pick up the main-branch, but also map the 1x1 kernel to 3x3 by padding zeros, and identity(also the input) should map into a 3x3 kernel also. You may reference the Fig4 in original paper.
![image](https://user-images.githubusercontent.com/7272659/117906350-53fa7d00-b307-11eb-9070-2654bfc42285.png)
This is why there are several function `get_equivalent_kernel_bias`, `_pad_1x1_to_3x3_tensor` and `_fuse_bn_tensor` in [DingXiaoH/RepVGG](https://github.com/DingXiaoH/RepVGG/blob/main/repvgg.py). If you only pick up the main-branch for inference, it will cause  the accuracy fall down.

~~~py
class RepVGGBlock(nn.Module):

     .........
    def get_equivalent_kernel_bias(self):
        kernel3x3, bias3x3 = self._fuse_bn_tensor(self.rbr_dense)
        kernel1x1, bias1x1 = self._fuse_bn_tensor(self.rbr_1x1)
        kernelid, biasid = self._fuse_bn_tensor(self.rbr_identity)
        return kernel3x3 + self._pad_1x1_to_3x3_tensor(kernel1x1) + kernelid, bias3x3 + bias1x1 + biasid

    def _pad_1x1_to_3x3_tensor(self, kernel1x1):
        if kernel1x1 is None:
            return 0
        else:
            return torch.nn.functional.pad(kernel1x1, [1,1,1,1])

    def _fuse_bn_tensor(self, branch):
        if branch is None:
            return 0, 0
        if isinstance(branch, nn.Sequential):
            kernel = branch.conv.weight
            running_mean = branch.bn.running_mean
            running_var = branch.bn.running_var
            gamma = branch.bn.weight
            beta = branch.bn.bias
            eps = branch.bn.eps
        else:
            assert isinstance(branch, nn.BatchNorm2d)
            if not hasattr(self, 'id_tensor'):
                input_dim = self.in_channels // self.groups
                kernel_value = np.zeros((self.in_channels, input_dim, 3, 3), dtype=np.float32)
                for i in range(self.in_channels):
                    kernel_value[i, i % input_dim, 1, 1] = 1
                self.id_tensor = torch.from_numpy(kernel_value).to(branch.weight.device)
            kernel = self.id_tensor
            running_mean = branch.running_mean
            running_var = branch.running_var
            gamma = branch.weight
            beta = branch.bias
            eps = branch.eps
        std = (running_var + eps).sqrt()
        t = (gamma / std).reshape(-1, 1, 1, 1)
        return kernel * t, beta - running_mean * gamma / std

    def switch_to_deploy(self):
        if hasattr(self, 'rbr_reparam'):
            return
        kernel, bias = self.get_equivalent_kernel_bias()
        self.rbr_reparam = nn.Conv2d(in_channels=self.rbr_dense.conv.in_channels, out_channels=self.rbr_dense.conv.out_channels,
                                     kernel_size=self.rbr_dense.conv.kernel_size, stride=self.rbr_dense.conv.stride,
                                     padding=self.rbr_dense.conv.padding, dilation=self.rbr_dense.conv.dilation, groups=self.rbr_dense.conv.groups, bias=True)
        self.rbr_reparam.weight.data = kernel
        self.rbr_reparam.bias.data = bias
        for para in self.parameters():
            para.detach_()
        self.__delattr__('rbr_dense')
        self.__delattr__('rbr_1x1')
        if hasattr(self, 'rbr_identity'):
            self.__delattr__('rbr_identity')
~~~

**Comment by YeongHyeon** [comment]
Thank you for the information.
Right now, it is difficult to reflect on your request.
I recommend you modify the source code through the pull request.
Thank you!
